types/recommendationrequest.go

```diff
@@ -0,0 +1,63 @@
+/*
+Copyright 2019 The MayaData Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package types
+
+import (
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// CStorClusterRecommendationRequest is a kubernetes custom
+// resource that represents configuration input or requirement
+// to create or scale one CStor pool.
+type CStorClusterRecommendationRequest struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+
+	Spec CStorClusterRecommendationRequestSpec `json:"spec"`
+}
+
+// CStorClusterRecommendationRequestSpec contains input or requirement
+// to create or scale one CStorpoolcluster.
+type CStorClusterRecommendationRequestSpec struct {
+	// Name represents name of the cspc.
+	Name string `json:"name"`
+	// Namespace represents namespace of the cspc.
+	Namespace string `json:"namespace"`
+	// PoolCapacity represents requested capacity for one pool
+	PoolCapacity resource.Quantity `json:"capacity"`
+	// DataConfig represents raid configuration for data devices.
+	DataConfig RaidGroupConfig `json:"dataConfig"`
+	// WriteCacheConfig represents raid configuration for write cache devices.
+	// If this field is nil then write cache is disabled.
+	WriteCacheConfig *RaidGroupConfig `json:"writeCacheConfig"`
+	// ReadCacheConfig represents raid configuration for read cache devices.
+	// If this field is nil then read cache is disabled.
+	ReadCacheConfig *RaidGroupConfig `json:"readCacheConfig"`
+}
+
+// RaidGroupConfig ...
+type RaidGroupConfig struct {
+	// Type is the raid group type
+	// Supported values are : stripe, mirror, raidz and raidz2
+	Type string `json:"type"`
+	// -- stripe DeviceCount = 1
+	// -- mirror DeviceCount = 2
+	// -- raidz DeviceCount = (2^n +1)
+	// -- raidz2 DeviceCount = (2^n +2)
+	DeviceCount int64 `json:"deviceCount"`
```

Comment:
How do we validate if the value specified can be validated with the actual block devices available?
Will users be given an option to choose BlockDevices or count of devices?
Does this API plan to support dynamic provisioning of BlockDevices via CSI (device count helps here)?

Reply:
> How do we validate if the value specified can be validated with the actual block devices available?
> Will users be given an option to choose BlockDevices or count of devices?
1. If the count of available block devices is less than this count then we will not get any recommendation.
2. This field is to decide how many devices will be in one raid group. If we take raidz and 9 device example then there are 2 options to create a single group of (8+1) devices or three groups of (2+1) devices.
> Does this API plan to support dynamic provisioning of BlockDevices via CSI (device count helps here)?
I think no. cc @avishnu